network/groups.go

```diff
@@ -83,72 +84,68 @@ func NewAdversaryGroups() (groups AdversaryGroups) {
+	g.updateGroupMana()
+
+	// Adversary nodes are taking indexes from the end, excluded randomly chosen nodes
+	advIndex := len(weightDistribution)
+	// weight distribution with adversary weights appended at the ned
+	newWeights := g.updateAdvIDAndWeights(advIndex, weightDistribution)
+
+	return newWeights
+}
+
+func (g *AdversaryGroups) updateAdvIDAndWeights(advIndex int, newWeights []uint64) []uint64 {
 	for groupIndex, group := range *g {
-		switch group.TargetMana {
-		case 100: // node has max
-			idx := selectMaxWeightIndex(ZIPFDistribution, usedWeights)
-			group.AddNodeID(idx, groupIndex)
-			group.GroupMana = float64(ZIPFDistribution[idx])
-			usedWeights[idx] = types.Void
-		case 0:
-			idx := selectMinWeightIndex(ZIPFDistribution, usedWeights)
-			group.AddNodeID(idx, groupIndex)
-			group.GroupMana = float64(ZIPFDistribution[idx])
-			usedWeights[idx] = types.Void
-		case -1:
-			// all weights are chosen randomly
-			for i, idx := range randomWeightIndex(ZIPFDistribution, len(*g)) {
-				(*g)[i].GroupMana = float64(ZIPFDistribution[idx])
-				(*g)[i].AddNodeID(idx, i)
-			}
-			return
-		default:
-			group.TargetMana = totalWeight * group.TargetMana / 100
-			manaPercentageGroupIds = append(manaPercentageGroupIds, groupIndex)
+		for i := 0; i < group.NodeCount; i++ {
+			group.AddNodeID(advIndex, groupIndex)
+			advIndex++
+			// append adversary weight at the end of weight distribution
+			nodeWeight := uint64(group.GroupMana / float64(group.NodeCount))
+			newWeights = append(newWeights, nodeWeight)
 		}
 	}
+	return newWeights
+}
 
-	// We define this because the adversary nodes' weights might not exactly occupy totalWeight * q
-	errorThreshold := totalWeight * config.AdversaryErrorThreshold
-	groupIndexQueue := queue.New(len(manaPercentageGroupIds))
-	// to choose indexes alternately and keep the balance between adversary groups
-	for _, groupIndex := range manaPercentageGroupIds {
-		groupIndexQueue.Offer(groupIndex)
-	}
-	for index, v := range ZIPFDistribution[config.AdversaryIndexStart:] {
-		for i := 0; i < groupIndexQueue.Size(); i++ {
-			elem, _ := groupIndexQueue.Poll()
-			groupIndex := elem.(int)
-			if (*g)[groupIndex].GroupMana < (*g)[groupIndex].TargetMana && (*g)[groupIndex].GroupMana+float64(v) < (*g)[groupIndex].TargetMana+errorThreshold {
-				(*g)[groupIndex].GroupMana += float64(v)
-				(*g)[groupIndex].AddNodeID(index+config.AdversaryIndexStart, groupIndex)
-				groupIndexQueue.Offer(groupIndex)
-				break
-			}
-			groupIndexQueue.Offer(groupIndex)
-		}
-	}
-	for groupIndex, group := range *g {
-		log.Infof("GroupId: %d, AdversaryType: %d, Node count: %d, q: %.5f, adNodeList: %v", groupIndex, group.AdversaryType, len(group.NodeIDs),
-			group.GroupMana/totalWeight, group.NodeIDs)
+func (g *AdversaryGroups) updateGroupMana() {
+	for _, group := range *g {
+		group.GroupMana = group.TargetManaPercentage * float64(config.NodesTotalWeight)
```

Comment:
```suggestion
		group.GroupMana = group.TargetManaPercentage * float64(config.NodesTotalWeight) / 100.0
```
Do we need to divide it by 100.0 here?